Add a task to run forbiddenapis using cli

alpar-t opened the pull request:
Add a task that offers an equivalent check to the forbidden APIs plugin,
but runs it using the forbiddenAPIs CLI instead.

This isn't wired into precommit first, and doesn't work for projects
that require specific signatures, etc. It's meant to show how this can
be used. The next step is to make a custom task type and configure it
based on the project extension from the pugin and make some minor
adjustments to some build scripts as we can't  bee 100% compatible with
that at least due to how additional signatures are passed.

Notes:
- there's no `--target` for the CLI version so we have to pass in
specific bundled signature names
- the cli task already wires to `runtimeJavaHome`
- no equivalent for `failOnUnsupportedJava = false` but doesn't seem to
be a problem. Tested with Java 8 to 11
- there's no way to pass additional signatures as URL, these will have
to be on the file system, and can't be resources on the cp unless we
rely on how forbiddenapis is implemented and mimic them as boundled
signatures.
- the average of 3 runs is 4% slower using the CLI for :server.
( `./gradlew clean :server:forbiddenApis` vs `./gradlew clean
:server:forbiddenApisCli`)
- up-to-date checks don't work on the cli task yet, that will happen
with the custom task.

See also: #31715

alpar-t committed:
Add a task to run forbiddenapis using cli

Add a task that offers an equivalent check to the forbidden APIs plugin,
but runs it using the forbiddenAPIs CLI instead.

This isn't wired into precommit first, and doesn't work for projects
that require specific signatures, etc. It's meant to show how this can
be used. The next step is to make a custom task type and configure it
based on the project extension from the pugin and make some minor
adjustments to some build scripts as we can't  bee 100% compatible with
that at least due to how additional signatures are passed.

Notes:
- there's no `--target` for the CLI version so we have to pass in
specific bundled signature names
- the cli task already wires to `runtimeJavaHome`
- no equivalent for `failOnUnsupportedJava = false` but doesn't seem to
be a problem. Tested with Java 8 to 11
- there's no way to pass additional signatures as URL, these will have
to be on the file system, and can't be resources on the cp unless we
rely on how forbiddenapis is implemented and mimic them as boundled
signatures.
- the average of 3 runs is 4% slower using the CLI for :server.
( `./gradlew clean :server:forbiddenApis` vs `./gradlew clean
:server:forbiddenApisCli`)
- up-to-date checks don't work on the cli task yet, that will happen
with the custom task.

See also: #31715

elasticmachine commented:
Pinging @elastic/es-core-infra

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy:
```diff
@@ -96,9 +103,62 @@ class PrecommitTasks {
+            forbiddenApisCliJar 'de.thetaphi:forbiddenapis:2.5'
+        }
+        Task forbiddenApisCli = project.tasks.create('forbiddenApisCli')
+        project.ext.getForbiddenSignaturesFile = { name -> project.getForbiddenSignaturesFile(name) }
```
Do we need this one? It looks circular.

atorok replied:
It's so we can call `getForbiddenSignaturesFile(name)` from a build script instead of `import ... ; PrecommitTasks.getForbiddenSignaturesFile(project, name)`

atorok replied:
I'll remove it for now, and add it back once it's actually used.

nik9000 commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy:
```diff
@@ -96,9 +103,62 @@ class PrecommitTasks {
+        return forbiddenApisCli
+    }
+
+    private static File getForbiddenSignaturesFile(project, String name) {
```
Could you add a type to the `project` parameter?

atorok replied:
I haven't found a good way to fight IDEA on this ... 
The refactoring don't add a type. Will change it.

nik9000 reviewed (changes requested):
This seems like a good start for forking this. I left some minor things.

alpar-t committed:
PR review

nik9000 reviewed (approved):
LGTM

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy:
```diff
@@ -96,9 +103,61 @@ class PrecommitTasks {
+                    executable = "${project.runtimeJavaHome}/bin/java"
+                    [
+                            'jdk-unsafe-1.8', 'jdk-deprecated-1.8', 'jdk-non-portable', 'jdk-system-out'
+                    ].forEach { args "-b", it }
```
I think this would be easier to read not using a foreach. Just call args directly with each argument.

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy:
```diff
@@ -96,9 +103,61 @@ class PrecommitTasks {
+                    [
+                            getForbiddenSignaturesFile(project, 'jdk-signatures'),
+                            getForbiddenSignaturesFile(project, 'es-all-signatures')
+                    ].forEach { args "-f", it }
```
Same here, the foreach just makes this harder to read.

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy:
```diff
@@ -96,9 +103,61 @@ class PrecommitTasks {
+    }
+
+    private static File getForbiddenSignaturesFile(Project project, String name) {
+        return project.rootProject.file('buildSrc/src/main/resources/forbidden/' + name + '.txt')
```
We can't assume this is accessible, since build-tools is used by external plugin developers. We need to load this from resources and copy to a local file.

atorok replied:
I wasn't aware we publish `group: 'org.elasticsearch.gradle', name: 'build-tools', version: '6.3.1'` and encourage others to use it.  Is this fully documented or more like plugin authors picked up on it ? I don't think we have documentation on any of the build plugins.  I think I broke this assumption recently with the naming conventions check, I'll take a look and see if we can add some tests for this.

atorok replied:
Naming conventions check is fine, it looks up the source jar of the relevant class so works as long as the jar is on the file system, the class doesn't have to be. I'll still add a test for this.

rjernst replied:
It's not "documented", but we have encouraged plugin authors to use it since we switched to gradle in 5.0. We have long needed a permanent home for plugin author documentation, but we probably need a broader reorg of the docs to enable that, as there are a few parts for plugin authors to consume eg how to build, how to test, core apis and services available, how to publish, etc.

rjernst reviewed (changes requested):
This looks good. The only critical thing I see is how the resource files are loaded.

alpar-t committed:
Add failing test to highlight paths issue

Fails with:
```
ERROR: IO problem while reading files with API signatures: java.io.FileNotFoundException: /home/alpar/work/elastic/elasticsearch/buildSrc/src/testKit/forbidden-apis/buildSrc/src/main/resources/forbidden/jdk-signatures.txt (No such file or directory)
```
because signature definitiions are accessed on the FS directly.

alpar-t committed:
Task skeleton with test failing due to missing implementation

alpar-t committed:
Task implementation and complete task

atorok commented:
Depends on  #32201 for completion.

alpar-t committed:
No more whitelist, reduce boilerplate

alpar-t committed:
Merge remote-tracking branch 'origin/master' into access-build-tools-resources

alpar-t committed:
Merge remote-tracking branch 'origin/master' into access-build-tools-resources

alpar-t committed:
remove unused enum

alpar-t committed:
Revert "Add failing test to highlight paths issue"

This reverts commit 60570d7ea231ba50ff7921432eb7d9be8b4c0f99.

alpar-t committed:
Merge branch 'access-build-tools-resources' into fix/forbidden-APIs-checks-jvm-31715

alpar-t committed:
Use the build resources task for running forbidden apis with teh CLI

alpar-t committed:
Merge branch 'master' into fix/forbidden-APIs-checks-jvm-31715

rjernst reviewed (approved):
LGTM aside from the naming I noted in #32201

alpar-t committed:
rename method to `copy`

atorok commented:
I changed the name to`copy`
